File: pc/internal/util/version_verifier/version_verifier.go
package version_verifier

import (
	"fmt"
	"log"
	"strconv"
	"strings"
)

// VerifyClientVersion expects client and server version strings in format "X.Y", where X and Y are integers.
//
// X is the major version and Y is the minor version.
// Returns nil if the major versions are the same.
func VerifyClientVersion(clientV, serverV string) error {
	cMajor, cMinor, err := parseVersion(clientV)
	if err != nil {
		return fmt.Errorf("client version parse failed: %w", err)
	}
	sMajor, sMinor, err := parseVersion(serverV)
	if err != nil {
		return fmt.Errorf("server version parse failed: %w", err)
	}

	if cMajor != sMajor {
		return fmt.Errorf("major version mismatch: client %d, server %d", cMajor, sMajor)
	}

	switch {
	case sMinor > cMinor:
		log.Println("Warning: server minor version is higher than client version")
	case sMinor < cMinor:
		log.Println("Warning: server minor version is lower than client version")
	}

	return nil
}

func parseVersion(version string) (major int, minor int, err error) {
	parts := strings.Split(version, ".")
	if len(parts) != 2 {
		return 0, 0, fmt.Errorf("invalid version format %q: expected major.minor", version)
	}
	major, err = strconv.Atoi(parts[0])
	if err != nil {
		return 0, 0, fmt.Errorf("invalid major version %q: %w", parts[0], err)
	}
	minor, err = strconv.Atoi(parts[1])
	if err != nil {
		return 0, 0, fmt.Errorf("invalid minor version %q: %w", parts[1], err)
	}
	return major, minor, nil
}

File: android/app/src/main/java/com/skymouse/skymouseclient/data/util/VersionVerifier.kt
package com.skymouse.skymouseclient.data.util


data class Version(val major: Int, val minor: Int) {
    companion object {
        /**
         * Expects a string in the format "X.Y", where X and Y are integers.
         * X is the major version and Y is the minor version.
         * Returns null if the string is not in the expected format.
         */
        fun parse(version: String): Version? {
            val parts = version.split(".")
            if (parts.size != 2) return null

            val major = parts[0].toIntOrNull() ?: return null
            val minor = parts[1].toIntOrNull() ?: return null

            return Version(major, minor)
        }
    }
}

sealed interface VersionVerificationResult {
    data object Valid : VersionVerificationResult
    data class Warning(val message: String) : VersionVerificationResult
    data class Mismatch(val reason: String) : VersionVerificationResult
}

object VersionVerifier {

    fun verify(clientVersion: String, serverVersion: String): VersionVerificationResult {
        val client = Version.parse(clientVersion)
            ?: return VersionVerificationResult.Mismatch("Client version parse failed: $clientVersion")

        val server = Version.parse(serverVersion)
            ?: return VersionVerificationResult.Mismatch("Server version parse failed: $serverVersion")

        if (client.major != server.major) {
            return VersionVerificationResult.Mismatch("Major version mismatch (Client: $clientVersion, Server: $serverVersion)")
        }

        return when {
            server.minor > client.minor ->
                VersionVerificationResult.Warning("Server minor version ($serverVersion) is higher than client ($clientVersion)\nClient update recommended")
            server.minor < client.minor ->
                VersionVerificationResult.Warning("Server minor version ($serverVersion) is lower than client ($clientVersion)\nServer update recommended")
            else ->
                VersionVerificationResult.Valid
        }
    }
}